relayer/relays/beacon/header/header.go

```diff
@@ -259,32 +258,162 @@ func (h *Header) SyncHeaders(ctx context.Context) error {
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
 
-	blockRoot := common.HexToHash(finalizedHeader.FinalizedHeader.ParentRoot.Hex())
-	prevSyncAggregate, err := h.syncer.GetSyncAggregateForSlot(uint64(finalizedHeader.FinalizedHeader.Slot + 1))
+func (h *Header) SyncHeaders(ctx context.Context, fromHeader, toHeader common.Hash, toHeaderSlot uint64, basicChannel, incentivizedChannel chan uint64) error {
+	fromHeaderUpdate, err := h.syncer.GetHeaderUpdate(fromHeader)
 	if err != nil {
 		return err
 	}
 
-	headerUpdate, err := h.SyncHeader(ctx, finalizedHeaderBlockRoot, prevSyncAggregate)
+	toHeaderUpdate, err := h.syncer.GetHeaderUpdate(toHeader)
 	if err != nil {
 		return err
 	}
 
-	prevSyncAggregate = headerUpdate.Block.Body.SyncAggregate
+	fromSlot := uint64(fromHeaderUpdate.Block.Slot)
+	toSlot := uint64(toHeaderUpdate.Block.Slot)
+	totalSlots := toSlot - fromSlot
 
-	for secondLastFinalizedHeader != blockRoot {
-		headerUpdate, err := h.SyncHeader(ctx, blockRoot, prevSyncAggregate)
+	log.WithFields(log.Fields{
+		"fromHeader": fromHeader,
+		"fromSlot":   fromSlot,
+		"fromEpoch":  h.syncer.ComputeEpochAtSlot(fromSlot),
+		"toHeader":   toHeader,
+		"toSlot":     toSlot,
+		"toEpoch":    h.syncer.ComputeEpochAtSlot(toSlot),
+		"totalSlots": totalSlots,
+	}).Info("starting to back-fill headers")
+
+	headersToSync := []syncer.HeaderUpdate{}
+
+	currentSlot := fromSlot + 1
+	for currentSlot <= toSlot {
+		epoch := h.syncer.ComputeEpochAtSlot(currentSlot)
+
+		// start of new epoch, sync headers of last epoch
+		if float64(epoch) >= (float64(currentSlot) / float64(h.syncer.SlotsInEpoch)) {
+			log.WithFields(log.Fields{
+				"epoch": h.syncer.ComputeEpochAtSlot(currentSlot) - 1,
+			}).Debug("syncing header in epoch")
+			for _, header := range headersToSync {
+				err := h.SyncHeader(ctx, header, toSlot-uint64(header.Block.Slot))
+				if err != nil {
+					return err
+				}
+			}
+
+			if len(headersToSync) > 0 {
+				err = h.sendLastBlockNumberMessage(ctx, uint64(headersToSync[len(headersToSync)-1].Block.Body.ExecutionPayload.BlockNumber), basicChannel, incentivizedChannel)
+				if err != nil {
+					return err
+				}
+			}
+
+			// new epoch, start with clean array
+			headersToSync = []syncer.HeaderUpdate{}
+		}
+
+		log.WithFields(log.Fields{
+			"slot": currentSlot,
+		}).Info("fetching header at slot")
+
+		headerUpdate, err := h.syncer.GetNextHeaderUpdateBySlot(currentSlot)
 		if err != nil {
 			return err
 		}
+		// To get the sync witness for the current synced header. This header
+		// will be used as the next update.
+		nextHeaderUpdate, err := h.syncer.GetNextHeaderUpdateBySlot(currentSlot + 1)
+		if err != nil {
+			return err
+		}
+
+		headerUpdate.SyncAggregate = nextHeaderUpdate.Block.Body.SyncAggregate
+
+		headersToSync = append(headersToSync, headerUpdate)
+		headerUpdate = nextHeaderUpdate
```

Comment:
`headerUpdate` isn't used after this, so we can remove it

```suggestion
		headersToSync = append(headersToSync, headerUpdate)
```

Reply:
It is used in the next iteration in the `for` loop, so it cannot be removed.

Reply:
But it'll be redefined on line 329 with the call to `GetNextHeaderUpdateBySlot`

Reply:
Ah good point! https://github.com/Snowfork/snowbridge/pull/696/commits/0034ef757d7ac5dea260fb038e2981f10ea87fc5